src/lib/user/actions/update-my-dog-registration.ts

```diff
@@ -0,0 +1,151 @@
+import { dbBegin, dbCommit, dbQuery, dbRelease } from "@/lib/data/db-utils";
+import { UserActor } from "../user-actor";
+import { MyDogRegistrationUpdate } from "../user-models";
+import { PoolClient } from "pg";
+import {
+  dbDeleteDogVetPreferences,
+  dbInsertDogVetPreference,
+} from "@/lib/data/db-dogs";
+
+type Result =
+  | "OK_UPDATED"
+  | "ERROR_REPORT_EXISTS"
+  | "ERROR_UNAUTHORIZED"
+  | "ERROR_MISSING_DOG"
+  | "FAILURE_DB_UPDATE";
+
+type Context = {
+  actor: UserActor;
+  update: MyDogRegistrationUpdate;
+};
+
+export async function updateMyDogRegistration(
+  actor: UserActor,
+  update: MyDogRegistrationUpdate,
+): Promise<Result> {
+  const { dbPool } = actor.getParams();
+  const ctx: Context = { actor, update };
+  const conn = await dbPool.connect();
+  try {
+    await dbBegin(conn);
+    const ownershipCheck = await checkOwnership(conn, ctx);
+    if (ownershipCheck !== "OK") {
+      return ownershipCheck;
+    }
+    const reportCheck = await checkExistingReport(conn, ctx);
+    if (reportCheck !== "OK") {
+      return reportCheck;
+    }
+    const updateFields = await updateDogFields(conn, ctx);
+    if (updateFields !== "OK") {
+      return updateFields;
+    }
+    updateVetPreference(conn, ctx);
```

Comment:
In the event where the call to updateVetPreference results in an exception, the effects of updateDogFields should be rolled back. Hence, the finally block we should call dbRollback before dbRelease. This will ensure that any execution flow that did not arrive at dbCommit would have no side-effects.

Reply:
Fixed